File: project/main/spkmeans.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../include/DDG.h"
#include "../include/eigengap.h"
#include "../include/jacobi.h"
#include "../include/kmeans.h"
#include "../include/LNORM.h"
#include "../include/parse_file.h"
#include "../include/WAM.h"
#include "../include/logger.h"
#include "../include/calculate_centroids.h"
#include <string.h>
#include <stdlib.h>

typedef enum {
	_SPK,
	_WAM,
	_DDG,
	_LNORM,
	_JACOBI
} goal;

typedef struct {
	goal goal;
	int K;
	char* filename;
} args;

args parse_cmd(int argc, char* argv[]) {
	args args;
	char* goal_str;
	int K;

	if (argc != 4) log_err("Wrong number of arguments. Correct format is: ./spkeams <goal> <K> <filename>\n");

	goal_str = argv[1];

	if (strcmp(goal_str, "spk") == 0) {
		args.goal = _SPK;
	} else if (strcmp(goal_str, "wam") == 0) {
		args.goal = _WAM;
	} else if (strcmp(goal_str, "ddg") == 0) {
		args.goal = _DDG;
	} else if (strcmp(goal_str, "lnorm") == 0) {
		args.goal = _LNORM;
	} else if (strcmp(goal_str, "jacobi") == 0) {
		args.goal = _JACOBI;
	} else log_err("BAD GOAL: %s\n", goal_str);

	K = atoi(argv[2]);

	if (K < 0) log_err("BAD K: %d; K must be non-negative.\n", K);

	args.K = K;

	args.filename = argv[3];

	return args;
}

void do_spkmeans(matrix* input, int K) {
	matrix *W, *D, *Lnorm, *T, *centroids;
	vectors_values_pair pair;
	vectors_k_pair verctors_pair;
	int* centroids_arr;
	int new_k;

	W = WAM(input);
	D = DDG(W);
	Lnorm = LNORM(W, D);
	pair = jacobi(Lnorm);
	verctors_pair = eigengap_heuristic(pair, K);
	T = verctors_pair.vectors;
	new_k = verctors_pair.k;
	centroids_arr = kmeans(T, new_k);
	centroids = calculate_centroids(centroids_arr, input, new_k);
	matrix_print(centroids);

	matrix_free(W);
	matrix_free(D);
	matrix_free(pair.eigenvectors);
	matrix_free(T);
	matrix_free(centroids);
	free(pair.eigenvalues);
	free(centroids_arr);
}

void do_wam(matrix* input) {
	matrix* res;

	res = WAM(input);
	matrix_print(res);
	matrix_free(res);
}

void do_ddg(matrix* input) {
	matrix *W, *res;

	W = WAM(input);
	res = DDG(W);
	matrix_print(res);
	matrix_free(W);
	matrix_free(res);
}

void do_lnorm(matrix* input) {
	matrix *res, *W, *D;

	W = WAM(input);
	D = DDG(W);
	res = LNORM(W, D);
	matrix_print(res);
	matrix_free(W);
	matrix_free(D);
	matrix_free(res);
}

void do_jacobi(matrix* input) {
	vectors_values_pair pair;
	if (input->n != input->m) log_err(
			"Jacobi matrix input should be symetrical.\n"); //TODO: make this better.

	pair = jacobi(input);

	if (!pair.eigenvectors) {
		printf("Jacobi algorithm did not converge after maximum iterations.");
		return;
	}

	matrix_print(pair.eigenvectors);
	eigenvectors_free(pair);
}

int main(int argc, char* argv[]) {
	args args;
	matrix* input;

	args = parse_cmd(argc, argv);
	input = read_csv(args.filename);
	if (args.K >= input->m) log_err(
		"BAD K: %d; K must be less than vectors count.\n", args.K);

	switch (args.goal) {
		case _SPK:
			do_spkmeans(input, args.K);
			break;
		case _WAM:
			do_wam(input);
			break;
		case _DDG:
			do_ddg(input);
			break;
		case _LNORM:
			do_lnorm(input);
			break;
		case _JACOBI:
			do_jacobi(input);
			break;
	}

	matrix_free(input);
	return 0;
}

File: project/python/spkmeans.py
import argparse, sys, os
import spkmeans as c_api
import numpy as np
import pandas as pd

np.random.seed(0)

###############################
# Parse command line arguments
###############################

def parse_arguments():
	parser = argparse.ArgumentParser()
	parser.add_argument('k', type=int, help="Number of clusters to create")
	parser.add_argument('goal', type=str, help="requested algorithm to preform")
	parser.add_argument('file_name', type=str, help="File name")

	args = parser.parse_args()
	if args.k < 0:
		print("K must be non-negative")
		sys.exit()

	if args.goal not in ['spk', 'wam', 'ddg', 'lnorm', 'jacobi']:
		print("Please enter a valid goal")
		sys.exit()

	return (args.k, args.goal, args.file_name)

########
# Utils
########

def print_val(val, end = "\n"):
	if abs(val) < 0.00005:
		val = 0
	print(f"{val:.4f}", end=end)

def print_vector(vec):
	[print_val(val, end=',') for val in vec[:-1]]
	print_val(vec[-1])


def print_matrix(mat, dim, vec_num):
	matrix = [mat[i:i + dim] for i in range(0, vec_num * dim, dim)]
	for row in matrix:
		print_vector(row)

##########################
# Algo 1.1 from Exercise 2
##########################

# Returns the distance between two vectors.
def dist(vec1, vec2):
	return sum(np.square(vec1 - vec2))


# Returns the distance from vector to closest centroid.
def closest_dist(centroids, vector):
	D = dist(vector[:-2], centroids.iloc[0].to_numpy()[:-2])
	for index, centroid in centroids.iterrows():
		tmp_val = dist(vector[:-2], centroid.to_numpy()[:-2])
		if (tmp_val <= D):
			D = tmp_val
	return D


def print_chosen_centroids(centroids):
	print(str(centroids.index.tolist()).strip("[]").replace(" ", ""))


# Returns a single python list of all vectors, where k centroids are first.
def format_vectors(centroids, vectors):
	vectors = vectors.drop(['Dist', 'Prob'], axis=1)
	centroids = centroids.drop(['Dist', 'Prob'], axis=1)
	vectors = vectors.drop([i for i in centroids.index])
	vectors = pd.concat([centroids, vectors])
	return vectors.to_numpy().flatten().tolist()


# Algo 1.1
def smart_centroids(vectors, K):
	N, d = vectors.shape
	if N <= K:
		return (None, None, None)

	vectors['Dist'] = vectors['Prob'] = 0.0
	centroids = vectors.iloc[[np.random.choice(N)]]

	Z = 1
	while (Z < K):
		for i in range(N):
			vectors['Dist'][i] = closest_dist(centroids, vectors.iloc[i].to_numpy())
		sumd = sum(vectors['Dist'])
		for i in range(N):
			vectors['Prob'][i] = vectors['Dist'][i] / sumd
		Z += 1
		row_num = np.random.choice(N, p=vectors['Prob'])
		centroids = centroids.append(vectors.iloc[row_num])

	print_chosen_centroids(centroids)
	num_arr = format_vectors(centroids, vectors)
	return num_arr

########
# C API
########

def do_wam(mat, dim, vec_num):
	try:
		func_name = 'WAM'
		W = c_api.WAM(mat, dim, vec_num)
	except ValueError as e:
		print(f"{func_name} method returned an error: {e}")
		sys.exit()
	
	print_matrix(W, vec_num, vec_num)


def do_ddg(mat, dim, vec_num):
	try:
		func_name = 'WAM'
		W = c_api.WAM(mat, dim, vec_num)
		func_name = 'DDG'
		D = c_api.DDG(W, vec_num)

	except ValueError as e:
		print(f"{func_name} method returned an error: {e}")
		sys.exit()

	print_matrix(D, vec_num, vec_num)


def do_Lnorm(mat, dim, vec_num):
	try:
		func_name = 'WAM'
		W = c_api.WAM(mat, dim, vec_num)
		func_name = 'DDG'
		D = c_api.DDG(W, vec_num)
		func_name = 'LNORM'
		L = c_api.LNORM(W, D, vec_num)

	except ValueError as e:
		print(f"{func_name} method returned an error: {e}")
		sys.exit()

	print_matrix(L, vec_num, vec_num)


def do_jacobi(mat, dim):
	try:
		eigenvectors, eigenvalues = c_api.jacobi(mat, dim)
	except TimeoutError:
		print("Jacobi method reached maximum iterations with no convergence.")
		sys.exit()
	except ValueError as e:
		print(f"Jacobi method returned an error: {e}")
		sys.exit()

	print_vector(eigenvalues)
	print_matrix(eigenvectors, dim, dim)


def do_spk(mat, dim, vec_num, k):
	try:
		func_name = 'WAM'
		W = c_api.WAM(mat, dim, vec_num)
		print('W')
		print_matrix(W, vec_num, vec_num)

		func_name = 'DDG'
		D = c_api.DDG(W, vec_num)

		func_name = 'LNORM'
		L = c_api.LNORM(W, D, vec_num)
		print('L')
		print_matrix(L, vec_num, vec_num)

		func_name = 'jacobi'
		eigenvectors, eigenvalues = c_api.jacobi(L, vec_num)
		print("Eigenvalues:")
		print_vector(eigenvalues)

		func_name = 'eigengap'
		T, k = c_api.eigengap_heuristic(eigenvectors, eigenvalues, k)
		matrix = [T[i:i + k] for i in range(0, vec_num * k, k)]
		data = pd.DataFrame(matrix)

		func_name = 'smart_centroids'
		T = smart_centroids(data, k)

		func_name = 'kmeans'
		index_list = c_api.kmeans(T, k, vec_num, k)

		func_name = 'calculate_centroids'
		centroids = c_api.calculate_centroids(index_list, mat, dim, vec_num, k)

	except ValueError as e:
		print(f"{func_name} method returned an error: {e}")
		sys.exit()

	print_matrix(centroids, dim,k)

#######
# Main
#######

if __name__ == "__main__":

	k, GOAL, FILENAME = parse_arguments()

	if not os.path.isfile(FILENAME):
		raise FileNotFoundError(f"No such file: {FILENAME}")

	data = pd.read_csv(FILENAME, header=None)
	dim = data.columns.size
	vec_num = data.shape[0]
	mat = data.to_numpy().flatten().tolist()

	if k >= vec_num:
		raise ValueError(f"BAD K: {k}; must be less than vectors count.")

	if GOAL == 'wam':
		do_wam(mat, dim, vec_num)
	elif GOAL == 'ddg':
		do_ddg(mat, dim, vec_num)
	elif GOAL == 'lnorm':
		do_Lnorm(mat, dim, vec_num)
	elif GOAL == 'jacobi':
		do_jacobi(mat, dim)
	elif GOAL == 'spk':
		do_spk(mat, dim, vec_num, k)